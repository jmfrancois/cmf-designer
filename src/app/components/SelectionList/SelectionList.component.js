import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { cmfConnect, Inject } from '@talend/react-cmf';
import Immutable from 'immutable';

import theme from './SelectionList.scss';

class SelectionList extends React.Component {
	constructor(props) {
		super(props);
		this.state = { opened: true };
		this.onClick = this.onClick.bind(this);
		this.onAddClick = this.onAddClick.bind(this);
		this.onClickToggle = this.onClickToggle.bind(this);
	}

	onClickToggle() {
		this.setState(({ opened }) => ({ opened: !opened }));
	}

	onClick(event, item) {
		const id = item.get('id');
		this.props.dispatch({
			type: SelectionList.ACTION_TYPE_SELECT_ITEM,
			componentId: this.props.componentId,
			id,
		});
	}
	onAddClick() {
		this.props.dispatch({
			type: SelectionList.ACTION_TYPE_ADD_ITEM,
			componentId: this.props.componentId,
		});
	}

	render() {
		return (
			<React.Fragment>
				<div className={theme.header}>
					<button className="btn btn-link" onClick={this.onClickToggle}>
						<Inject
							component="Icon"
							componentId={this.state.opened ? 'selection-list-opened' : 'selection-list-closed'}
						/>
					</button>
					<h2>{this.props.title}</h2>
					<button className={classNames('btn btn-primary', theme.add)} onClick={this.onAddClick}>
						<Inject component="Icon" componentId="selection-list-add" />
					</button>
				</div>
				{this.state.opened && (
					<div className="list-group">
						{this.props.items.map((item, index) => (
							<button
								key={index}
								className="list-group-item"
								onClick={event => this.onClick(event, item)}
							>
								{item.get('name')}
							</button>
						))}
					</div>
				)}
			</React.Fragment>
		);
	}
}

SelectionList.propTypes = {
	title: PropTypes.string,
	items: PropTypes.array,
	onClick: PropTypes.func,
	activeId: PropTypes.string,
	...cmfConnect.propTypes,
};
SelectionList.defaultProps = {
	items: [],
};
SelectionList.displayName = 'SelectionList';
SelectionList.ACTION_TYPE_SELECT_ITEM = 'SelectionList#selectItem';
SelectionList.ACTION_TYPE_ADD_ITEM = 'SelectionList#addItem';

export default cmfConnect({
	defaultState: new Immutable.Map(),
})(SelectionList);
