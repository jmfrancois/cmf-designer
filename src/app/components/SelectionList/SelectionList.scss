.header {
	position: relative;
	display: flex;
	align-items: center;
	padding: $padding-normal 0;

	h2 {
		margin: 0;
	}

	.add {
		position: absolute;
		right: 0;
	}
}
