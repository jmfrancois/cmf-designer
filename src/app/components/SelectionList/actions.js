export function onClickSelectionListItem(event, data) {
	if (event.persist) {
		event.persist();
	}
	return {
		type: 'CLICK_SELECTION_LIST_ITEM',
		id: data.item.get('id'),
		event,
	};
}

