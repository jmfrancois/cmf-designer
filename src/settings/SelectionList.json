{
  "props": {
    "SelectionList#components": {
      "title": "Components",
      "itemsExpression": {
        "id": "cmf.collections.get",
        "args": ["components.components", []]
      }
    },
    "SelectionList#props": {
      "title": "Props",
      "itemsExpression": {
        "id": "cmf.collections.get",
        "args": ["props.data", []]
      }
    }
  }
}
