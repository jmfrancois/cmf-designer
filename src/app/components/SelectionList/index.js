import SelectionList from './SelectionList.component';
import actions from './actions';

SelectionList.actions = actions;
export default SelectionList;
