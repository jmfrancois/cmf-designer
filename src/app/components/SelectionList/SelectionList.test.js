import React from 'react';
import { shallow } from 'enzyme';

import SelectionList from './SelectionList.component';

describe('SelectionList', () => {
	it('should render', () => {
		const wrapper = shallow(<SelectionList.WrappedComponent />);
		expect(wrapper.getElement()).toMatchSnapshot();
	});
});
