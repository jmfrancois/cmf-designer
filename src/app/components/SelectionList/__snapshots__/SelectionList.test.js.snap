// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectionList should render 1`] = `
<div>
  <h2>
    <button
      className="btn btn-default btn-xs"
      onClick={[Function]}
    >
      <Inject
        component="Icon"
        componentId="selection-list-opened"
      />
    </button>
    <span />
    <button
      className="btn btn-primary btn-xs"
      onClick={[Function]}
    >
      <Inject
        component="Icon"
        componentId="selection-list-add"
      />
    </button>
  </h2>
  <div
    className="list-group"
  >
    
  </div>
</div>
`;
